core/services/pipeline/runner.go

```diff
@@ -236,10 +236,16 @@ func init() {
 // results, even after context cancellation.
 func overtimeContext(ctx context.Context) (context.Context, context.CancelFunc) {
 	if d, ok := ctx.Deadline(); ok {
-		// extend deadline
-		return context.WithDeadline(context.WithoutCancel(ctx), d.Add(overtime))
-	}
-	// remove cancellation
+		// We do not use context.WithDeadline/Timeout in order to prevent the monitor hook from logging noisily, since
+		// we expect and want these operations to use most of their allotted time.
+		// TODO replace with custom thresholds: https://smartcontract-it.atlassian.net/browse/BCF-3252
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithCancel(context.WithoutCancel(ctx))
+		t := time.AfterFunc(time.Until(d.Add(overtime)), cancel)
+		stop := context.AfterFunc(ctx, func() { t.Stop() })
+		return ctx, func() { cancel(); stop() }
+	}
+	// do not propagate cancellation in any case
```

Comment:
As I understand this code, instead of returning a context with a deadline of `d + overtime`, we're returning a context which technically has no automatic deadline on it, but which gets manually cancelled after `d + overtime` has elapsed.  Is that accurate?

If that's the case, then I don't think this fixes the underlying problem.  When the query goes past `d + overtime` it will get aborted either way, resulting in CRIT errors.  This will silence the ERROR and WARN's, but as far as I can tell it won't change anything about how much time the actual query has to execute.

Reply:
Do you mean that there is a fundamental problem, or just that the other fix to recognize cancelled vs. deadlinemissed is required too? https://github.com/smartcontractkit/chainlink-common/pull/530

Reply:
I think it may not matter in this case, because we do in fact want to know about the critical ones which did not meet their deadline. If this happened before, then we just did not see it at all. The question is whether we are now short-circuiting things which could have still returned results _after_ the (extended) deadline is missed.

Reply:
Actually, wait... when does anything call cancel on the returned child context? Won't this just allow the query to run forever... because there is no deadline and the cancel of its parent will not propagate to the child?

Also, for readability:  might be better to use a new variable for the `ctx` that's being returned instead of overwriting the input param

Reply:
> Actually, wait... when does anything call cancel on the returned child context? Won't this just allow the query to run forever... because there is no deadline and the cancel of its parent will not propagate to the child?

Only in the case where there was no fixed deadline. The default 10s timeout will still be applied

Reply:
Oh, right! Okay, so it has no deadline at this point but... later the default timeout gets added by the sql layer, got it.

Still a bit unsure of when cancel() gets called on this child context... does that happen automatically once the default timeout on the query is hit?

Reply:
Actually, this will call it after `d + overtime` has elapsed, right?
```
t := time.AfterFunc(time.Until(d.Add(overtime)), cancel)
```
If I understand that right, looks like the query will still get aborted early, not wait for the full 10s

Reply:
Yes, that case will be aborted and we will log, but if the monitor logs it is actually meaningful, unlike the lower threshold warnings and errors for just 10%/50% of timeout.

Reply:
Then I'm back to my original point which is that, this doesn't do anything to prevent the queries from timing out... which has been happening in 2.12 but not in 2.11. All it does is silence ERROR's and WARN's when the query is close to timing out.

(Unless the other PR you linked to changes that somehow--still looking at that, but it doesn't seem like it would.)

Reply:
Oh, I see... the other PR silences the CRIT's in the case where it's a cancel rather than a deadline.  But is silencing them what we want to do? Is the idea that... they may happen from time to time, but hopefully will succeed later after being re-tried?

Reply:
No the idea is that cancellation is fundamentally different from missing a deadline. It is not inherently problematic to cancel things, and abandoning work quietly is the correct thing to do. We don't have any additional information at the level of the monitor. If the cancellation is truly a problem, then that is the concern and responsibility of the upstream code to deal with.